### PR TITLE
FINERACT-2119: Address jobs by name

### PR DESCRIPTION
**Pull request by marta-jankovics**
## Description

[Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).](https://issues.apache.org/jira/browse/FINERACT-2119)

New field on job table: 
short_name VARCHAR(8) NNULL, Unique
Example: ACC_AATR (Add Accrual Transactions) - first part is like a namespace, second part is specific for the job
 
namespaces:
LA: Loan Account
SA: Savings Account
SH: Shares
ACC: Accounting
STI: Standing Instruction
EM: Email sending
SMS: Sms sending
BDT: Business Date
COM: Commands
EXE: External Events

New API endpoints:
These API endpoints were added by the story above. 
GET /v1/jobs/short-name/{shortName}
GET /v1/jobs/short-name/{shortName}/runhistory
POST /v1/jobs/short-name/{shortName}
PUT /v1/jobs/short-name/{shortName}

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/resources/db/changelog/tenant/parts/0146_add_final_constraints.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied. See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.1.xsd">
+    <changeSet author="fineract" id="1" context="postgresql">
```
Comment (adamsaghy) [inline fineract-provider/src/main/resources/db/changelog/tenant/parts/0146_add_final_constraints.xml:25]: Would you mind to write an sql check here to enforce not-null constraints if there is no entry in the job table that does not have short name filled?

**Review by adamsaghy — APPROVED** [review]
LGTM

File: custom/acme/loan/job/src/main/resources/db/custom-changelog/0001_acme_loan_job.xml [hunk 2]
```diff
@@ -42,4 +42,10 @@
             <column name="is_mismatched_job" valueBoolean="true"/>
         </insert>
     </changeSet>
+    <changeSet author="acme" id="2">
+        <update tableName="job">
+            <column name="short_name" value="ACM_NOOP"/>
+            <where>name='Acme Noop Job'</where>
+        </update>
+    </changeSet>
```
Comment (Zeyad2003) [inline custom/acme/loan/job/src/main/resources/db/custom-changelog/0001_acme_loan_job.xml:50]: @adamsaghy @marta-jankovics 

I am currently working on a custom module in another PR (#3994), and when I rebased the branch with the `develop` branch and ran the custom Docker image, I encountered the following error:

```sql
fineract-fineract-1  | Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'tenantDatabaseUpgradeService' defined in URL [jar:file:/app/libs/fineract-provider-0.1.0-SNAPSHOT-plain.jar!/org/apache/fineract/infrastructure/core/service/migration/TenantDatabaseUpgradeService.class]: liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset db/custom-changelog/0001_acme_loan_job.xml::1::acme:
fineract-fineract-1  |      Reason: liquibase.exception.DatabaseException: ERROR: null value in column "short_name" of relation "job" violates not-null constraint
fineract-fineract-1  |   Detail: Failing row contains (40, Acme Noop Job, Acme Noop Job, 0 1 0 1/1 * ? *, 2024-08-29 15:30:47.958143, 5, null, null, null, Acme Noop Job _ DEFAULT, null, f, f, t, 0, f, 1, t, null). [Failed SQL: (0) INSERT INTO public.job (name, display_name, cron_expression, create_time, task_priority, group_name, previous_run_start_time, job_key, initializing_errorlog, is_active, currently_running, updates_allowed, scheduler_group, is_misfired, node_id, is_mismatched_job) VALUES ('Acme Noop Job', 'Acme Noop Job', '0 1 0 1/1 * ? *', NOW(), 5, NULL, NULL, 'Acme Noop Job _ DEFAULT', NULL, FALSE, FALSE, TRUE, 0, FALSE, 1, TRUE)]
```

Could you please explain why this occurred and how I can resolve it?

> [!NOTE]
> I am running [`docker-compose-api-v3.yml`](https://github.com/Zeyad2003/fineract/blob/c63762763f40cfb8390bdbf53a9eb6c425f63e0c/docker-compose-api-v3.yml) which uses **Postgres** image

Comment (adamsaghy) [reply custom/acme/loan/job/src/main/resources/db/custom-changelog/0001_acme_loan_job.xml:50]: I guess the problem might be the order of execution of the liquibase scripts. @marta-jankovics What do you think?

Comment (marta-jankovics) [reply custom/acme/loan/job/src/main/resources/db/custom-changelog/0001_acme_loan_job.xml:50]: 0001_acme_loan_job.xml
should run before 
0146_add_final_constraints.xm
make sure that in
db/changelog/db.changelog-master.xml
your module changeling-master.xml is listed before
tenant/final-changelog-tenant.xml

Comment (marta-jankovics) [reply custom/acme/loan/job/src/main/resources/db/custom-changelog/0001_acme_loan_job.xml:50]: The problem is that scripts marked with custom_changelog run after the normal run was finished. This is not correct. I'll fix this with a a small PR today.
See https://github.com/apache/fineract/pull/4041

**Pull request merged**